src/ray/raylet/node_manager.cc

```diff
@@ -1918,7 +1918,11 @@ void NodeManager::HandleTaskBlocked(const std::shared_ptr<LocalClientConnection>
     // blocked task matches the one assigned to the worker, then mark the
     // worker as blocked. This temporarily releases any resources that the
     // worker holds while it is blocked.
+    // TODO: For direct calls, the current task ID provided by the client may
+    // not match the assigned task ID. It's possible to get resource deadlock
+    // in this situation.
     if (!worker->IsBlocked() && current_task_id == worker->GetAssignedTaskId()) {
+      RAY_LOG(DEBUG) << "Releasing resources for task " << current_task_id;
```

Comment:
I believe we always route those to HandleDirectCallTaskBlocked, so we should never hit this path for direct calls.

Reply:
Oh hmm that's odd. I definitely saw these log messages come up when I was testing, and I have your commit merged into my branch.

Reply:
I think what's happening is that it will first call `Get` on the in-memory store, and that will correctly block/unblock the worker, but then in this case we also have to fetch the objects through plasma. The plasma store's `Get` will then go through this code path.

Is that expected behavior? It seems like the DirectCallTaskBlocked/Unblocked calls should go before and after the core worker's `Get`, not just the in-memory store's `Get`.

Reply:
Yeah that seems bad, we should modify the plasma store to also use the direct block unblock calls. We should only do this if the object does not already exist, to avoid any unnecessary IPCs.

Reply:
Okay, I'll just move the TODO for now then.

Reply:
Sounds good. I'll make a fix.